Store subpartition template in catalog with transformed values

l-wang opened the pull request:
Previously, when storing gpPartitionDefinition as subpartition
template in the `gp_partition_template` catalog table, we store
information as received from parser. It was not the best practice
because:

- Serialize and deserialize for A_Const had to be written because we
  are writing non-transformed values to disk.
- ALTER TABLE ADD PARTITION with a template has to transform the
  template again, which is duplicate effort that can be avoided.
- Having the non-transformed values complicates the implementation of
  the pg_get_expr() rule for gpPartitionDefinition which is used by
  pg_dump.

In this PR we start storing the transformed
gpPartitionDefinition in gp_partition_template.

Now with a dedicated parser step that transforms
gpPartitionDefinition, we now use the transformed
gpPartitionDefinition to generate the CREATE TABLE statements for the
child partitions. We also took the chance to separate the list of
ENCODING clauses from the list of gpPartitionDefElems.

ashwinstar commented:
seems there are pipeline failures due to the change.

soumyadeep2007 commented on src/backend/commands/tablecmds_gp.c:
```diff
@@ -1369,7 +1369,7 @@ ATExecGPPartCmds(Relation origrel, AlterTableCmd *cmd)
 				 */
 				generatePartitions(firstchildoid, templateDef, NULL, cmd->queryString,
-								   NIL, NULL, NULL, true);
+								   NIL, NULL, NULL, true, false);
```
Isn't it possible to completely decouple the act of storing the partition template from the act of generating partitions? `generatePartitions()` should only generate partitions.

Why not have something like this:
```c
generatePartitions(firstchildoid, templateDef, NULL, cmd->queryString,
								   NIL, NULL, NULL, true, false);
if (isSubTemplate)
    StoreGpPartitionTemplate(..);
```

Inside `generatePartitions()`, `StoreGpPartitionTemplate` is only called if `subPartSpec->gpPartDef->fromCatalog` is true.

I noticed that `subPartSpec->gpPartDef->fromCatalog` is only set to true for an `ATExecGPPartCmds`, which is part of alter table processing. 

In fact there is already a callsite like this in `ATExecGPPartCmds`:

```
/*
 * call generatePartitions() using first child as partition to
 * add partitions, just to validate the subpartition template,
 * if anything wrong it will error out.
 */
generatePartitions(firstchildoid, templateDef, NULL, cmd->queryString,
				   NIL, NULL, NULL, true);
table_close(firstrel, AccessShareLock);
StoreGpPartitionTemplate(topParentrelid, level, templateDef);
```

If we do this, we can get rid of `fromCatalog` entirely from `GpPartitionDefinition` and also do the much needed decoupling. Thoughts?

soumyadeep2007 replied:
It would also be nice if we can take `GpPartitionDefinition->encClauses` out as it seems to be dead code.

soumyadeep2007 replied:
I am fine with not squashing. Thanks for the explanation.

soumyadeep2007 reviewed (commented):
Reviewed up to the 2nd commit so far. Will continue tomorrow.

soumyadeep2007 commented on src/backend/parser/parse_partition_gp.c:
```diff
@@ -1440,88 +1231,339 @@ convert_exclusive_start_inclusive_end(Const *constval, Oid part_col_typid, int32
+		listdatums = lappend(listdatums, linitial(thisvalue));
 	}
+	boundspec->listdatums = listdatums;
+	boundspec->location = -1;
```
Out of scope for this PR but noting it down:
I find it weird that we don't set `boundspec->location` to `GpPartDefElem->location`, like we do with range partitioning. 
Same for the default partition in `generateDefaultPartition`.

l-wang replied:
Good point. I'm noting this down and see what I can do in a future PR.

soumyadeep2007 reviewed (commented):
Great work and thanks for adding so much coverage!

Mostly have minor code suggestions and a few test suggestions for some of the elog ERRORs that was moved around by the refactor (and for which no pre-existing tests exist):

* number of EVERY values should cover all partition key columns
* VALUES specification with more than one column not allowed

huansong commented on src/backend/parser/parse_partition_gp.c:
```diff
@@ -1537,119 +1579,224 @@ generatePartitions(Oid parentrelid, GpPartitionDefinition *gpPartSpec,
 			else
 				partDefElems = lappend(partDefElems, elem);
 		}
+		else
+		{
+			Assert(IsA(newnode, ColumnReferenceStorageDirective));
+			encClauses = lappend(encClauses, newnode);
+		}
```
+1
Though it could be tricky to fix because of how encoding clauses end up mixing with normal `GpPartDefElem` in the grammar. If so we probably should restore the above FIXME and fix it in another day.

huansong reviewed (approved):
LGTM!

l-wang commented on src/backend/parser/parse_partition_gp.c:
```diff
@@ -1537,119 +1579,224 @@ generatePartitions(Oid parentrelid, GpPartitionDefinition *gpPartSpec,
+				case PARTITION_STRATEGY_LIST:
+					transformGpPartDefElemWithListSpec(pstate, parentrel, elem);
+					break;
+				default: elog(ERROR, "Not supported partition strategy");
```
@huansong here's I think:
- we don't want to add support for GPDB-style hash partition syntax because we are encouraging users to start using the PG-style syntax.
- Since we got hash partition functionality "for free" from Postgres, I see no reason to ban it if users use PG-style syntax.

So IMO I don't see an action item for hash partition for now.

l-wang replied:
@soumyadeep2007 I'm keeping this as is for now. It guards nonsense commands like this:
```
create table foo (a int, b int) partition by hash (b) (partition p1 start (1) end (2));
```
We didn't have test coverage for this. Adding one.

l-wang commented on src/backend/parser/parse_partition_gp.c:
```diff
@@ -1537,119 +1579,224 @@ generatePartitions(Oid parentrelid, GpPartitionDefinition *gpPartSpec,
+generatePartitions(Oid parentrelid, GpPartitionDefinition *gpPartSpec,
+				   PartitionSpec *subPartSpec, const char *queryString,
+				   List *parentoptions, const char *parentaccessmethod,
+				   List *parentattenc, bool forvalidationonly)
```
It's not really dead at the moment, it is used to decided whether or not we deduce implicit range partition bounds. See comments in `generatePartitions()`. Essentially there are slight differences among CREATE TABLE, ALTER TABLE ADD PARTITION and ALTER TABLE SET SUBPARTITION TEMPLATE commands. However, I also dislike the fact that we are skipping validation for SET SUBPARTITION TEMPLATE, we might end up storing a bogus template. Let me see what I can do there.
```c
	/*
	 * GPDB range partition
	 *
	 * Validate and maybe update range partitions bound here instead of in
	 * check_new_partition_bound(), because we need to modify the lower or upper
	 * bounds for implicit START/END.
	 *
	 * Need to skip this step forvalidationonly -- which is called by SET
	 * SUBPARTITION TEMPLATE. Reason is deduceImplicitRangeBounds() assumes
	 * for ADD PARTITION, only one partition is being added if missing START
	 * or END specification. While that's true for ADD PARTITION, it's not
	 * while setting template.
	 */
	if (hasImplicitRangeBounds && !forvalidationonly)
		deduceImplicitRangeBounds(pstate, parentrel, result);
```

l-wang commented:
@soumyadeep2007 @huansong Thank you so much for reviewing! I addressed all of the feedback and also added two new commits:
- [Don't skip validation for SET SUBPARTITION TEMPLATE](https://github.com/greenplum-db/gpdb/pull/14211/commits/2420d219c05704b45f45dbf96f810cf12c7d2eea)
- [Add more test coverage for SET SUBPARTITION TEMPLATE](https://github.com/greenplum-db/gpdb/pull/14211/commits/46ff526117c0dd2e825aea7da60ad0418a66ce1c)

Please take another look if have chance. I plan to squash all of the refactor related commits but keep the above two commits separate.

soumyadeep2007 commented on src/backend/parser/parse_partition_gp.c:
```diff
@@ -801,8 +802,37 @@ generateRangePartitions(ParseState *pstate,
+	 *
+	 * create table z (a int, b int, c int) partition by range(b, c);
+	 * create table z1 partition of z for values from (10, 10) TO (20, 20);
+	 */
```
Lets add this as errhint.

soumyadeep2007 commented on src/backend/parser/parse_partition_gp.c:
```diff
@@ -887,18 +918,41 @@ generateListPartition(Relation parentrel,
+	 * create table z (a int, b int, c int) partition by range(b, c);
+	 * create table z1 partition of z for values from (10, 10) TO (20, 20);
+	 */
+
```
Same here, this can be an errhint!

l-wang replied:
This comment was copy pasted from the other function but I get what the author was trying to say. Unfortunately we don't have an existing alternative syntax for multi-column list partitioning. We have discussions regarding different workarounds, but I'm not sure how to state that in the errhint. Since this is a well known issue I'm going to remove the FIXME altogether.

soumyadeep2007 commented on src/backend/parser/parse_partition_gp.c:
```diff
@@ -1565,6 +1619,12 @@ transformGpPartitionDefinition(Oid parentrelid, const char *queryString,
+	Assert(partkey != NULL);
+	if (list_length(partkey->partexprs) > 0)
+		ereport(ERROR,
+				(errcode(ERRCODE_SYNTAX_ERROR),
```
Lets provide errhint suggesting the alternative syntax wherever we have a case where we have some new syntax that can succeed where old syntax fails (or vice-versa)

l-wang replied:
We wouldn't really get here due to syntax error, but really due to mixed usage of syntax. I plan to have a more holistic go though of error messages and error hints as separate work. For now I changed the error code to `ERRCODE_INVALID_OBJECT_DEFINITION`.

soumyadeep2007 reviewed (approved):
Looks good, thanks for adding all of the additional coverage!
Just have one suggestion about errhints.
